activerecord/lib/active_record/delay_touching/state.rb

```diff
@@ -0,0 +1,71 @@
+module ActiveRecord
+  # = Active Record Delay Touching
+  module DelayTouching
+
+    # Tracking of the touch state. This class has no class-level data, so you can
+    # store per-thread instances in thread-local variables.
+    class State # :nodoc:
```

Comment:
Here's the commit I was thinking of: https://github.com/rails/rails/commit/dc3f25c8a5aa64de9225f11498a389a2d31e880a

The code Aaron worked on has similar characteristics to this one. He let the `Scope` class handle a single level and then move back and forth using `parent` and `new` (an instance method) respectively.